libs/core/langchain_core/indexing/base.py

```diff
@@ -1,8 +1,414 @@
+    
+    Please note that deleting an ID that does not exist is **NOT** considered a failure.
+    """
+
+
+class BaseIndex(Generic[T]):
+    """A collection of items that can be queried.
+
+    This indexing interface is designed to be a generic abstraction for storing and
+    querying content that has an ID and metadata associated with it.
+
+    The interface is designed to be agnostic to the underlying implementation of the
+    indexing system.
+
+    The interface is designed to support the following operations:
+
+    1. Storing content in the index.
+    2. Retrieving content by ID.
+    3. Querying the content based on the metadata associated with the content.
+
+    The implementation is **NOT** responsible for supporting search queries
+    against the content itself! Such a responsibility is left for more specialized
+    interfaces like the VectorStore.
+
+    While strongly encouraged, implementations are not required to support
+    querying based on metadata. Such implementations override the `get_by_query`
+    and `delete_by_query` methods to raise a NotImplementedError.
+
+    .. versionadded:: 0.2.15
+    """
+
+    # Developer guidelines:
+    # Do not override streaming_upsert!
+    # This interface will likely be extended in the future with additional support
+    # to deal with failures and retries.
+    @beta(message="Added in 0.2.11. The API is subject to change.")
+    def streaming_upsert(
+        self, items: Iterable[T], /, batch_size: int, **kwargs: Any
+    ) -> Iterator[UpsertResponse]:
+        """Upsert items in a streaming fashion.
+
+        Args:
+            items: Iterable of content to add to the vectorstore.
+            batch_size: The size of each batch to upsert.
+            **kwargs: Additional keyword arguments.
+                kwargs should only include parameters that are common to all
+                items. (e.g., timeout for indexing, retry policy, etc.)
+                kwargs should not include ids to avoid ambiguous semantics.
+                Instead the ID should be provided as part of the item.
+
+        .. versionadded:: 0.2.11
+        """
+        # The default implementation of this method breaks the input into
+        # batches of size `batch_size` and calls the `upsert` method on each batch.
+        # Subclasses can override this method to provide a more efficient
+        # implementation.
+        for item_batch in batch_iterate(batch_size, items):
+            yield self.upsert(item_batch, **kwargs)
+
+    @abc.abstractmethod
+    @beta(message="Added in 0.2.15. The API is subject to change.")
+    def upsert(self, items: Sequence[T], /, **kwargs: Any) -> UpsertResponse:
+        """Upsert items into the index.
+
+        The upsert functionality should utilize the ID field of the content object
+        if it is provided. If the ID is not provided, the upsert method is free
+        to generate an ID for the content.
```

Comment:
Can you provide more strict guidance here? It's common to stick to [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119) terms.

Specifically -- is this "SHOULD" (recommended) or "MAY" (Optional) behavior?